File: backend/app/routers/contacts.py
"""Contacts API endpoints."""

from typing import Annotated, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.deps import get_current_user
from app.database import get_db
from app.schemas.contact import (
    ContactCreate,
    ContactUpdate,
    ContactResponse,
    ContactListResponse,
    ContactListMeta,
)
from app.services.contact import ContactService, contact_to_response_dict


router = APIRouter(prefix="/contacts", tags=["contacts"])


@router.get("", response_model=ContactListResponse)
async def list_contacts(
    current_user: Annotated[dict, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
    q: Optional[str] = Query(
        None,
        description="Search query for contact name (uses pg_trgm)",
        max_length=100,
    ),
    limit: int = Query(20, ge=1, le=100, description="Maximum number of contacts"),
    offset: int = Query(0, ge=0, description="Number of contacts to skip"),
):
    """
    Get paginated list of contacts with optional search.

    - **q**: Search contacts by name (uses pg_trgm for fuzzy matching)
    - **limit**: Maximum number of contacts to return (1-100, default 20)
    - **offset**: Number of contacts to skip for pagination

    Phone numbers and emails are decrypted in the response.
    """
    service = ContactService(db)
    contacts, total = await service.get_list(q=q, limit=limit, offset=offset)

    return ContactListResponse(
        data=[
            ContactResponse(**contact_to_response_dict(contact))
            for contact in contacts
        ],
        meta=ContactListMeta(total=total, limit=limit, offset=offset),
    )


@router.get("/{contact_id}", response_model=ContactResponse)
async def get_contact(
    contact_id: int,
    current_user: Annotated[dict, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """
    Get a single contact by ID.

    Phone number and email are decrypted in the response.
    """
    service = ContactService(db)
    contact = await service.get_by_id(contact_id)

    if not contact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contact not found",
        )

    return ContactResponse(**contact_to_response_dict(contact))


@router.post("", response_model=ContactResponse, status_code=status.HTTP_201_CREATED)
async def create_contact(
    data: ContactCreate,
    current_user: Annotated[dict, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """
    Create a new contact.

    - **name**: Required contact name
    - **phone**: Phone number (will be encrypted at rest)
    - **email**: Email address (will be encrypted at rest)
    - **organization**: Organization/company name
    - **position/role**: Contact's role or position
    - **notes**: Additional notes
    """
    service = ContactService(db)
    contact = await service.create(data)

    return ContactResponse(**contact_to_response_dict(contact))


@router.patch("/{contact_id}", response_model=ContactResponse)
async def update_contact(
    contact_id: int,
    data: ContactUpdate,
    current_user: Annotated[dict, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """
    Update an existing contact.

    Only provided fields are updated. Omit fields to leave them unchanged.
    """
    service = ContactService(db)
    contact = await service.get_by_id(contact_id)

    if not contact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contact not found",
        )

    updated_contact = await service.update(contact, data)

    return ContactResponse(**contact_to_response_dict(updated_contact))


@router.delete("/{contact_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_contact(
    contact_id: int,
    current_user: Annotated[dict, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """
    Soft delete a contact.

    The contact is not permanently removed but marked as deleted.
    Associated meeting attendee records are preserved.
    """
    service = ContactService(db)
    contact = await service.get_by_id(contact_id)

    if not contact:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contact not found",
        )

    await service.delete(contact)

    return None

File: frontend/src/lib/stores/contacts.ts
/**
 * Contacts Store - Contact state management for meeting attendees
 */
import { writable, derived } from 'svelte/store';
import { api } from '$lib/api';

export interface Contact {
	id: number;
	name: string;
	role: string | null;
	organization: string | null;
	phone?: string;
	email?: string;
	created_at: string;
	updated_at: string;
}

export interface ContactCreate {
	name: string;
	role?: string;
	organization?: string;
	phone?: string;
	email?: string;
}

export interface ContactUpdate {
	name?: string;
	role?: string;
	organization?: string;
	phone?: string;
	email?: string;
}

interface ContactsState {
	contacts: Contact[];
	selectedContact: Contact | null;
	isLoading: boolean;
	error: string | null;
	searchQuery: string;
	total: number;
}

function createContactsStore() {
	const { subscribe, set, update } = writable<ContactsState>({
		contacts: [],
		selectedContact: null,
		isLoading: false,
		error: null,
		searchQuery: '',
		total: 0
	});

	return {
		subscribe,

		async fetchContacts(query?: string, limit = 50, offset = 0): Promise<void> {
			update((state) => ({ ...state, isLoading: true, error: null }));

			try {
				const params: Record<string, string | number | undefined> = {
					limit,
					offset
				};
				if (query) {
					params.q = query;
				}

				const response = await api.get<{ data: Contact[]; meta: { total: number } }>(
					'/contacts',
					params
				);

				update((state) => ({
					...state,
					contacts: response.data,
					total: response.meta.total,
					isLoading: false,
					searchQuery: query || ''
				}));
			} catch (error) {
				const message = error instanceof Error ? error.message : 'Failed to fetch contacts';
				update((state) => ({ ...state, isLoading: false, error: message }));
			}
		},

		async fetchContact(id: number): Promise<Contact | null> {
			update((state) => ({ ...state, isLoading: true, error: null }));

			try {
				const response = await api.get<{ data: Contact }>(`/contacts/${id}`);
				update((state) => ({
					...state,
					selectedContact: response.data,
					isLoading: false
				}));
				return response.data;
			} catch (error) {
				const message = error instanceof Error ? error.message : 'Failed to fetch contact';
				update((state) => ({ ...state, isLoading: false, error: message }));
				return null;
			}
		},

		async createContact(data: ContactCreate): Promise<Contact | null> {
			update((state) => ({ ...state, isLoading: true, error: null }));

			try {
				const response = await api.post<{ data: Contact }>('/contacts', data);
				update((state) => ({
					...state,
					contacts: [response.data, ...state.contacts],
					total: state.total + 1,
					isLoading: false
				}));
				return response.data;
			} catch (error) {
				const message = error instanceof Error ? error.message : 'Failed to create contact';
				update((state) => ({ ...state, isLoading: false, error: message }));
				return null;
			}
		},

		async updateContact(id: number, data: ContactUpdate): Promise<Contact | null> {
			update((state) => ({ ...state, isLoading: true, error: null }));

			try {
				const response = await api.patch<{ data: Contact }>(`/contacts/${id}`, data);
				update((state) => ({
					...state,
					contacts: state.contacts.map((c) => (c.id === id ? response.data : c)),
					selectedContact:
						state.selectedContact?.id === id ? response.data : state.selectedContact,
					isLoading: false
				}));
				return response.data;
			} catch (error) {
				const message = error instanceof Error ? error.message : 'Failed to update contact';
				update((state) => ({ ...state, isLoading: false, error: message }));
				return null;
			}
		},

		async deleteContact(id: number): Promise<boolean> {
			update((state) => ({ ...state, isLoading: true, error: null }));

			try {
				await api.delete(`/contacts/${id}`);
				update((state) => ({
					...state,
					contacts: state.contacts.filter((c) => c.id !== id),
					selectedContact: state.selectedContact?.id === id ? null : state.selectedContact,
					total: state.total - 1,
					isLoading: false
				}));
				return true;
			} catch (error) {
				const message = error instanceof Error ? error.message : 'Failed to delete contact';
				update((state) => ({ ...state, isLoading: false, error: message }));
				return false;
			}
		},

		setSelectedContact(contact: Contact | null): void {
			update((state) => ({ ...state, selectedContact: contact }));
		},

		clearError(): void {
			update((state) => ({ ...state, error: null }));
		},

		reset(): void {
			set({
				contacts: [],
				selectedContact: null,
				isLoading: false,
				error: null,
				searchQuery: '',
				total: 0
			});
		}
	};
}

export const contactsStore = createContactsStore();

// Derived stores for convenience
export const contacts = derived(contactsStore, ($store) => $store.contacts);
export const selectedContact = derived(contactsStore, ($store) => $store.selectedContact);
export const contactsLoading = derived(contactsStore, ($store) => $store.isLoading);
export const contactsError = derived(contactsStore, ($store) => $store.error);
